parquet-variant/src/main/java/org/apache/parquet/variant/VariantConverters.java

```diff
@@ -29,755 +29,720 @@
 
 import java.math.BigDecimal;
 import java.math.BigInteger;
+import java.nio.ByteBuffer;
 import java.util.HashSet;
-import java.util.List;
 import java.util.Set;
+import java.util.function.Consumer;
+import org.apache.parquet.Preconditions;
 import org.apache.parquet.column.Dictionary;
 import org.apache.parquet.io.api.Binary;
 import org.apache.parquet.io.api.Converter;
 import org.apache.parquet.io.api.GroupConverter;
 import org.apache.parquet.io.api.PrimitiveConverter;
 import org.apache.parquet.schema.GroupType;
 import org.apache.parquet.schema.LogicalTypeAnnotation;
+import org.apache.parquet.schema.LogicalTypeAnnotation.ListLogicalTypeAnnotation;
 import org.apache.parquet.schema.PrimitiveType;
 import org.apache.parquet.schema.Type;
 
-class VariantConverters {
+public class VariantConverters {
   // do not allow instantiating this class
-  private VariantConverters() {}
+  private VariantConverters() {
+  }
 
-  /**
-   * Converter for a shredded Variant containing a value and/or typed_value field: either a top-level
-   * Variant column, or a nested array element or object field. The top-level converter is handled
-   * by a subclass (VariantColumnConverter) that also reads metadata.
-   *
-   * Converters for the `value` and `typed_value` fields are implemented as nested classes in this
-   * class.
-   *
-   * All converters for a Variant column append their results to a VariantBuilder as values are read from Parquet.
-   *
-   * Values in `typed_value` are appended by the child converter. Values in `value` are stored by a
-   * child converter, but only appended when completing this group. Additionally, object fields are
-   * appended by the `typed_value` converter, but because residual values are stored in `value`, this
-   * converter is responsible for finalizing the object.
-   */
-  static class VariantElementConverter extends GroupConverter implements VariantConverter {
+  public interface ParentConverter<T extends VariantBuilder> {
+    void build(Consumer<T> buildConsumer);
+  }
 
-    private int valueIdx = -1;
-    private int typedValueIdx = -1;
-    protected Converter[] converters;
+  public static GroupConverter newVariantConverter(GroupType variantGroup, Consumer<ByteBuffer> metadata,
+                                                   ParentConverter<VariantBuilder> parent) {
+    ValueConverter converter = newValueConverter(variantGroup, parent);
 
-    // The following are only used if this is an object field.
-    private String objectFieldName = null;
-    private VariantConverter parent = null;
+    // if there is a metadata field, add a converter for it
+    if (variantGroup.containsField("metadata")) {
```

Comment:
There are two main classes that handle the `value` / `typed_value` group that are created by `newValueConverter`. Those are `ShreddedObjectConverter` to handle shredded fields and `ShreddedValueConverter` to handle shredded arrays, scalars, and for when `typed_value` is missing. Rather than create subclasses that also handle `metadata`, this adds a converter to the existing classes.